x-pack/legacy/plugins/apm/server/routes/settings/agent_configuration.ts

```diff
@@ -102,55 +103,47 @@ export const agentConfigurationAgentNameRoute = createRoute(() => ({
     const setup = await setupRequest(context, request);
     const { serviceName } = context.params.query;
     const agentName = await getAgentNameByService({ serviceName, setup });
-    return agentName;
+    return { agentName };
   }
 }));
 
-export const createAgentConfigurationRoute = createRoute(() => ({
-  method: 'POST',
-  path: '/api/apm/settings/agent-configuration/new',
-  params: {
-    body: agentPayloadRt
-  },
+export const createOrUpdateAgentConfigurationRoute = createRoute(() => ({
+  method: 'PUT',
+  path: '/api/apm/settings/agent-configuration',
   options: {
     tags: ['access:apm', 'access:apm_write']
   },
+  params: {
+    query: t.partial({ overwrite: jsonRt.pipe(t.boolean) }),
+    body: agentConfigurationIntakeRt
+  },
   handler: async ({ context, request }) => {
     const setup = await setupRequest(context, request);
-    const configuration = context.params.body;
+    const { body, query } = context.params;
 
-    // TODO: Remove logger. Only added temporarily to debug flaky test (https://github.com/elastic/kibana/issues/51764)
-    context.logger.info(
-      `Hitting: /api/apm/settings/agent-configuration/new with ${configuration.service.name}/${configuration.service.environment}`
-    );
-    const res = await createOrUpdateConfiguration({
-      configuration,
+    // if the config already exists, it is fetched and updated
+    // this is to avoid creating two configs with identical service params
+    const config = await findExactConfiguration({
+      service: body.service,
       setup
     });
-    context.logger.info(`Created agent configuration`);
 
-    return res;
-  }
-}));
+    // if the config exists ?overwrite=true is required
+    if (config && !query.overwrite) {
+      throw Boom.badRequest(
+        `A configuration already exists for "${body.service.name}/${body.service.environment}. Use ?overwrite=true to overwrite the existing configuration.`
+      );
+    }
+
+    context.logger.info(
+      `${config ? 'Updating' : 'Creating'} config ${body.service.name}/${
+        body.service.environment
+      }`
+    );
 
-export const updateAgentConfigurationRoute = createRoute(() => ({
-  method: 'PUT',
-  path: '/api/apm/settings/agent-configuration/{configurationId}',
-  options: {
-    tags: ['access:apm', 'access:apm_write']
-  },
-  params: {
-    path: t.type({
-      configurationId: t.string
-    }),
-    body: agentPayloadRt
-  },
-  handler: async ({ context, request }) => {
-    const setup = await setupRequest(context, request);
-    const { configurationId } = context.params.path;
     return await createOrUpdateConfiguration({
-      configurationId,
-      configuration: context.params.body,
+      configurationId: config?._id,
```

Comment:
Just a thought, what about using a serialized representation of service.name/environment for id ? Not sure if it's better, but then you don't need to search first, and you don't have a possible race condition either.

Reply:
> what about using a serialized representation of service.name/environment

Great idea. Will make deleting easier.

As for updating I'm not sure, since we'll still need to lookup whether the config exists if `overwrite` is false.

Should we maybe just get rid of the `overwrite` param?

Reply:
Should have made the comment here: https://github.com/elastic/kibana/pull/57767#discussion_r380656441